File: Top 100 Codes/06-Advanced Codes related to Arrays/14-Find all pairs on integer array whose sum is equal to given number/Findallpairsonintegerarraywhosesumisequaltogivennumber.cpp
#include <iostream>
#include <vector>
#include <map>
using namespace std;

// Function to find and print all pairs with a given sum
void findPairsWithSum(const vector<int>& arr, int targetSum) {
    map<int, int> numCount; // Map to store frequency of elements
    bool found = false;

    for (int i = 0; i < arr.size(); i++) {
        int current = arr[i];
        int complement = targetSum - current;

        // If complement exists in the map, print it
        if (numCount[complement] > 0) {
            cout << "Pair found: (" << complement << ", " << current << ")\n";
            found = true;
        }

        // Add current number to map
        numCount[current]++;
    }

    if (!found) {
        cout << "No pairs found with sum " << targetSum << ".\n";
    }
}

int main() {
    // Example usage
    vector<int> arr = {1, 5, 7, -1, 5};
    int targetSum = 6;

    cout << "Array: ";
    for (int i = 0; i < arr.size(); i++) {
        cout << arr[i] << " ";
    }
    cout << "\nTarget Sum: " << targetSum << endl;

    findPairsWithSum(arr, targetSum);

    return 0;
}

File: Top 100 Codes/06-Advanced Codes related to Arrays/14-Find all pairs on integer array whose sum is equal to given number/Findallpairsonintegerarraywhosesumisequaltogivennumber.py
def find_pairs_with_sum(arr, target_sum):
    num_count = {}  # Dictionary to store frequencies
    found = False

    for i in range(len(arr)):
        current = arr[i]
        complement = target_sum - current

        # Check if complement exists in dictionary
        if complement in num_count:
            for _ in range(num_count[complement]):
                print(f"Pair found: ({complement}, {current})")
            found = True

        # Add current number to dictionary
        if current in num_count:
            num_count[current] += 1
        else:
            num_count[current] = 1

    if not found:
        print(f"No pairs found with sum {target_sum}.")

# Example usage
arr = [1, 5, 7, -1, 5]
target_sum = 6

print("Array:", arr)
print("Target Sum:", target_sum)
find_pairs_with_sum(arr, target_sum)
